nncf/quantization/algorithms/bias_correction/algorithm.py

```diff
@@ -0,0 +1,534 @@
+        the sub-graph and the model without quantizer-dequantizer pair or fake quantize node.
+        - at the next step, we collect the new statistics for the next layer (that would be corrected) from
+        the sub-graph with the updated bias value on the current step;
+        - after the new statistics were collected, we drops the unnecessary statistics to reduce memory consumption;
+        - in the end, we correct all needed biases in the original model.
+
+    :param number_samples: The number of the samples for the statistics collection.
+    :param threshold: The magnitude threshold that regulates the application of the shift.
+    :param nncf_graph: NNCFGraph class for the algorithm.
+    """
+
+    def __init__(self, parameters: BiasCorrectionParameters) -> None:
+        """
+        :param parameters: The instance of the BiasCorrectionParameters.
+        """
+        super().__init__()
+        self.number_samples = max(np.int(parameters.number_samples * 0.2), 1)
+        self.threshold = parameters.threshold
+        self.nncf_graph = None
+        self._backend_entity = None
+        self._collected_stat_inputs = set()
+        self._fp_inputs = {}
+
+    @property
+    def available_backends(self) -> Dict[str, BackendType]:
+        return ALGO_BACKENDS.registry_dict
+
+    def _set_backend_entity(self, model: TModel) -> None:
+        """
+        Creates a helper class with a backed-specific logic of the algorithm.
+
+        :param model: Backend-specific input model.
+        """
+        model_backend = get_backend(model)
+        if model_backend == BackendType.ONNX:
+            from nncf.quantization.algorithms.bias_correction.onnx_backend import \
+                ONNXBCAlgoBackend
+            self._backend_entity = ONNXBCAlgoBackend()
+        else:
+            raise RuntimeError('Cannot return backend-specific entity'
+                               'because {} is not supported!'.format(model_backend))
+
+    def _apply(self,
+               model: TModel,
+               statistic_points: Optional[StatisticPointsContainer] = None,
+               dataset: Optional[Dataset] = None) -> TModel:
+
+        self._set_backend_entity(model)
+        main_transformations_layout = TransformationLayout()
+        main_model_transformer = self._backend_entity.model_transformer(model)
+
+        model_copy = deepcopy(model)
+        model_copy = self._remove_fq_from_inputs(model_copy)
+        nncf_graph = NNCFGraphFactory.create(model_copy)
+        subgraphs_data = self._fill_subgraphs_data(nncf_graph)
+
+        for node_name, subgraph_data in subgraphs_data.items():
+            node = nncf_graph.get_node_by_name(node_name)
+
+            if not self._is_node_with_bias(node):
+                nncf_logger.debug('Skipping node {} because there is no bias'.format(node_name))
+                continue
+            if not self._backend_entity.is_quantized_weights(node, model):
```

Comment:
Could you briefly explain why you can not determinate it in the common part of the algorithm?

Reply:
This is because the weights quantize using different approaches from backend to backend. 
For example, Quantize-Dequantize pairs for ONNX and FakeQuantize nodes for IR. To determine quantized weights or not, we should watch (or move) through the backend-specific subgraph.

Reply:
Essentially, is_quantize_weights function finds quantization nodes on the constant branch. If the search algorithm knows which metatypes defines quantization nodes than the logic can be common. This case is similar as `QuantizerPropagationSolver`.